src/Compilers/CSharp/Portable/Lowering/Instrumentation/ModuleCancellationInstrumenter.cs

```diff
@@ -0,0 +1,198 @@
+    {
+        Previous.InterceptCallAndAdjustArguments(ref method, ref receiver, ref arguments, ref argumentRefKindsOpt, invokedAsExtensionMethod, nameSyntax);
+
+        if (arguments is [.., { Type: { } lastArgumentType } lastArgument] &&
+            (argumentRefKindsOpt.IsDefault || argumentRefKindsOpt is [.., RefKind.None]) &&
+            lastArgumentType.Equals(_throwMethod.ContainingType, TypeCompareKind.ConsiderEverything))
+        {
+            // The last argument is a CancellationToken. Replace it with the module-level token.
+            // Keep the previous expression so that side-effects are preserved.
+            arguments = [.. arguments[0..^1], _factory.Sequence([lastArgument], _factory.ModuleCancellationToken())];
+        }
+        else if (FindOverloadWithCancellationToken(method) is { } cancellableOverload)
+        {
+            // The method being invoked does not have a CancellationToken as the last parameter, but there is an overload that does.
+            // Invoke the other overload instead and pass in module-level token. 
+            method = cancellableOverload;
+            arguments = [.. arguments, _factory.ModuleCancellationToken()];
+            argumentRefKindsOpt = argumentRefKindsOpt.IsDefault ? default : [.. argumentRefKindsOpt, RefKind.None];
+        }
+    }
+
+    /// <summary>
+    /// Find an overload whose last parameter is <see cref="CancellationToken"/> and
+    /// the parameter types and ref kinds of all other parameters match those of <paramref name="method"/>.
+    /// </summary>
+    private MethodSymbol? FindOverloadWithCancellationToken(MethodSymbol method)
+    {
+        // It's unlikely that real-world APIs have overloads that differ in dynamic,
+        // but if they do avoid selecting them since their intended use might differ.
+        //
+        // Similarly, we also only consider overloads with the same visibility.
+        // We could potentially allow the visibility to differ,
+        // but we'd need to check if the overload is accessible at the call site.
+
+        const TypeCompareKind TypeComparisonKind = TypeCompareKind.CLRSignatureCompareOptions & ~TypeCompareKind.IgnoreDynamic;
+
+        foreach (var member in method.ContainingType.GetMembers(method.Name))
+        {
+            if (member.IsStatic == method.IsStatic &&
+                member.MetadataVisibility == method.MetadataVisibility &&
+                member is MethodSymbol { Parameters: [.., { RefKind: RefKind.None, Type: { } lastParamType }] parametersWithCancellationToken } overload &&
+                overload.Arity == method.Arity &&
+                method.Parameters.Length == parametersWithCancellationToken.Length - 1 &&
+                lastParamType.Equals(_throwMethod.ContainingType, TypeCompareKind.ConsiderEverything))
+            {
+                var typeMap = (method.Arity > 0) ? new TypeMap(overload.TypeParameters, method.TypeArgumentsWithAnnotations, allowAlpha: true) : null;
```

Comment:
>allowAlpha: true [](http://example.com/codeflow?start=125&length=16)

Why is this necessary, we are not alpha renaming anything here.

Reply:
For this scenario:
```C#
class D<T>
{
    public static void G<S>(T a) {}
    public static void G<S>(S a, CancellationToken token) {}
}

class C
{
    void F(CancellationToken token)
    {
        D<int>.G<int>(1);
    }
}
```

Reply:
That is likely an artifact of the fact that we are not looking at definitions. Once that is changed, let's take a look at this issue again.